[ Mathematical Functions ] Added the function named age_cbrt()

Zahid07 opened the pull request:
Takes an agtype input and returns the cube root.

Added Regression test cases for the same.

Useage:  

`SELECT * FROM age_cbrt(agtype_in('125'));`

```
Expected Output:
age_cbrt
----------
 5.0
(1 row)
```

Useage:
`SELECT * FROM age_cbrt(125);`

```
Expected Output:
ERROR:  cbrt() invalid input type
```

Zahid07 committed:
Added the function named age_cbrt()
Takes an agtype input and returns the cube root.

Added Regression test cases for the same.

JoshInnis commented on regress/expected/expr.out:
```diff
@@ -5117,6 +5117,59 @@ ERROR:  schema "contains" does not exist
+--------
+ 
+(1 row)
+
```
add a null call without the typecast

Zahid07 replied:
For the pg_catalog.cbrt or the ag_catalog.age_cbrt

JoshInnis replied:
ag_catalog In general. Thats what I care about

Zahid07 committed:
Changed Stable to Immutable as the cbrt Function
does not lookup the database for an answer rather it calculates an
answer at runtime.

Zahid07 committed:
Added a null Test Case for age_cbrt() function.

JoshInnis commented on src/backend/utils/adt/agtype.c:
```diff
@@ -7817,6 +7817,95 @@ Datum age_sqrt(PG_FUNCTION_ARGS)
+    float8 result;
+
+    /* Check If input is NULL */
+    if (agtype_in == NULL)
```
because of line 7826, this is redundant

Zahid07 replied:
Done

JoshInnis commented on age--1.1.0.sql:
```diff
@@ -3800,6 +3800,13 @@ IMMUTABLE
+RETURNS agtype
+LANGUAGE c
+IMMUTABLE
+PARALLEL SAFE
```
Add a line: RETURNS NULL ON NULL INPUT

Zahid07 replied:
Done

JoshInnis commented on src/backend/utils/adt/agtype.c:
```diff
@@ -7817,6 +7817,95 @@ Datum age_sqrt(PG_FUNCTION_ARGS)
+{
+
+    // check if input is agtype
+    if (PG_ARGISNULL(0))
```
By adding the line in age--1.1.0.sql this is unecesssary

Zahid07 replied:
Done

JoshInnis commented on src/backend/utils/adt/agtype.c:
```diff
@@ -7817,6 +7817,95 @@ Datum age_sqrt(PG_FUNCTION_ARGS)
+
+
+    /* Check if input is scalar type or not */
+    if (!agtype_extract_scalar(&agtype_in->root, &agtv) ||
```
Checkout get_float_compatible_arg, it will simplify all this, generally speaking, you should not do any casting in the function itself. Especially here, because Postgres changed float8in_internal_null in pg15 and this will add extra steps to PG15 support

Zahid07 replied:
For, get_float_compatible_arg I would need to change the argument from agtype to 'variadic"any" ', to cater the paremeters for get_float_compatible_arg .

Zahid07 replied:
Well I can use get_float_compatible_arg, and when I get the `type`, I can make a check there if the input is agtype or not.

Zahid07 replied:
Done

Zahid07 committed:
Removed Redundant NULL Calls
Added Null Check in the age--1.1.0.sql

Zahid07 committed:
Changed the way the cbrt function works to be more
efficient.

JoshInnis commented on age--1.1.0.sql:
```diff
@@ -3800,6 +3800,14 @@ IMMUTABLE
 PARALLEL SAFE
 AS 'MODULE_PATHNAME';
 
+CREATE FUNCTION ag_catalog.age_cbrt(variadic "any")
```
Change: 'variadic "any"' to 'agtype'

Zahid07 replied:
Done

Zahid07 committed:
Changed the argument type

JoshInnis commented on src/backend/utils/adt/agtype.c:
```diff
@@ -7817,6 +7817,48 @@ Datum age_sqrt(PG_FUNCTION_ARGS)
+     * and numeric for the input
+     */
+
+    input = get_float_compatible_arg(args[0], types[0], "cbrt", &is_null);
```
you can just pass the orginal argument: AG_GET_ARG_AGTYPE_P(0) instead of args[0] and delete a lot of the code above this line

JoshInnis replied:
the secornd arg (types[0]) can be AGTYPEOID

Zahid07 replied:
Okay

JoshInnis commented on src/backend/utils/adt/agtype.c:
```diff
@@ -7817,6 +7817,48 @@ Datum age_sqrt(PG_FUNCTION_ARGS)
+    bool is_null = true;
+
+    /* extract argument values */
+    nargs = extract_variadic_args(fcinfo, 0, true, &args, &types, &nulls);
```
everything here to the next comment can be deleted

Zahid07 replied:
Done

Zahid07 committed:
Optimized the cbrt Function
